subflow.go

```diff
@@ -0,0 +1,286 @@
+	ch := make(chan *frame)
+	r := byteReader{Reader: sf.conn}
+	go func() {
+		defer close(ch)
+		for {
+			var sz, fn uint64
+			var n int
+			sz, err = ReadVarInt(r)
+			if err != nil {
+				sf.close()
+				return
+			}
+			fn, err = ReadVarInt(r)
+			if err != nil {
+				sf.close()
+				return
+			}
+			if sz == 0 {
+				sf.gotACK(fn)
+				continue
+			}
+			log.Tracef("got frame %d from %s with %d bytes", fn, sf.to, sz)
+			buf := pool.Get(int(sz))
+			n, err = io.ReadFull(r, buf)
+			if err != nil {
+				pool.Put(buf)
+				sf.close()
+				return
+			}
+			if n != int(sz) {
+				panic(fmt.Sprintf("expect to read %d bytes, read %d", sz, n))
+			}
+			sf.ack(fn)
+			ch <- &frame{fn: fn, bytes: buf}
+			sf.tracker.OnRecv(sz)
+			select {
+			case <-sf.chClose:
+				return
+			default:
+				// continue
+			}
+		}
+	}()
+	probeTimer := time.NewTimer(randomize(probeInterval))
+	for {
+		select {
+		case frame := <-ch:
+			if frame == nil {
+				return
+			}
+			sf.mpc.recvQueue.add(frame)
+			if !probeTimer.Stop() {
+				<-probeTimer.C
+			}
+			probeTimer.Reset(randomize(probeInterval))
+		case <-probeTimer.C:
+			go sf.probe()
+			probeTimer.Reset(randomize(probeInterval))
+		}
+	}
+}
+
+func (sf *subflow) sendLoop() {
+	for {
+		select {
+		case <-sf.chClose:
+			return
+		case frame := <-sf.sendQueue:
+			sf.addPendingAck(frame)
+			n, err := sf.conn.Write(frame.buf)
+			if err != nil {
+				log.Debugf("failed to write frame %d to %s: %v", frame.fn, sf.to, err)
+				// TODO: For temporary errors, maybe send the subflow to the
+				// back of the line instead of closing it.
+				sf.close()
+				if frame.isDataFrame() {
+					go sf.mpc.retransmit(frame)
+				}
+				return
+			}
+			if n != len(frame.buf) {
+				panic(fmt.Sprintf("expect to write %d bytes on %s, written %d", len(frame.buf), sf.to, n))
+			}
+			if !frame.isDataFrame() {
+				frame.release()
+				continue
+			}
+			log.Tracef("done writing frame %d with %d bytes via %s", frame.fn, frame.sz, sf.to)
+			if frame.retransmissions == 0 {
+				sf.tracker.OnSent(frame.sz)
+			} else {
+				sf.tracker.OnRetransmit(frame.sz)
+			}
+			d := sf.retransTimer()
+			time.AfterFunc(d, func() {
+				if sf.isPendingAck(frame.fn) {
+					// No ack means the subflow fails or has a longer RTT
+					sf.updateRTT(d)
+					sf.mpc.retransmit(frame)
+				} else {
+					// It is ok to release buffer here as the frame will never
+					// be retransmitted again.
+					frame.release()
+				}
+			})
+		}
+	}
+}
+
+func (sf *subflow) ack(fn uint64) {
+	select {
+	case <-sf.chClose:
+	case sf.sendQueue <- composeFrame(fn, nil):
+	}
+}
+
+func (sf *subflow) gotACK(fn uint64) {
+	if fn == frameTypePing {
+		log.Tracef("pong to %s", sf.to)
+		sf.ack(frameTypePong)
+		return
+	}
+	sf.muPendingAcks.Lock()
+	defer sf.muPendingAcks.Unlock()
+	e := sf.pendingAcks.Front()
+	if e == nil {
+		panic("unsolicited ack")
+	}
+	pending := e.Value.(pendingAck)
+	if pending.fn != fn {
+		panic("unsolicited ack")
```

Comment:
This panic (and the one above on line 193) means that someone could crash the server or a client by sending an unsolicited ACK.  I realize that (a) that's pretty unlikely and (b) we could use something like panicwrap.  Still, seems less than ideal.  Perhaps we could just log these?

Reply:
This won't happen unless someone changes the multipath implementation *and* has the valid config (or has access to the proxy to panic the client), which in reality will only be us. If it logs the error and continues running, `pendingAcks` would grow unboundedly, the "realtime" RTT would keep increasing, or some other undefined behavior would happen, so it feels to me that the best we can do is to fail soundly. We can always patch the server to either prevent itself from crashing, or fix bugs causing clients crashing, and we have facility to monitor panics.

Reply:
Okay, yeah that's fair.  I was actually thinking that `multipath` would be the lowest layer, with pluggable transports layered on top.  I'm realizing now that multiplath needs to be layered on top of the pluggable transports or the traffic will stand out.  That also makes it much less likely (in most cases impossible) that an attacker could inject packets into the stream.